ethereum/sidecar/server.go

```diff
@@ -428,6 +537,342 @@ func (s *Server) startGRPCServer(
+
+	for _, confirmedEvent := range confirmedEvents {
+		confirmedUnlockSequence := confirmedEvent.UnlockSequenceNumber
+
+		// Only consider an event as attested if its unlock sequence still
+		// remains among confirmed unlocks.
+		isConfirmed, err := s.bridgeContract.ConfirmedUnlocks(confirmedUnlockSequence)
+		if err != nil {
+			return nil, fmt.Errorf(
+				"failed to check if AssetsUnlocked event remains confirmed: [%w]",
+				err,
+			)
+		}
+
+		if isConfirmed {
+			attestedUnlockSequences[confirmedUnlockSequence.String()] = true
+		}
+	}
+
+	// If the event's unlock sequence is not among the attested sequences
+	// consider the event unattested.
+	unattestedEvents := []bridgetypes.AssetsUnlockedEvent{}
+	for _, event := range inputEvents {
+		if !attestedUnlockSequences[event.UnlockSequence.String()] {
+			unattestedEvents = append(unattestedEvents, event)
+		}
+	}
+
+	// Sort the unattested events by their unlock sequences in ascending order.
+	sort.Slice(unattestedEvents, func(i, j int) bool {
+		return unattestedEvents[i].UnlockSequence.
+			LT(unattestedEvents[j].UnlockSequence)
+	})
+
+	return unattestedEvents, nil
+}
+
+func (s *Server) fetchNewAssetsUnlockedEvents(ctx context.Context) error {
+	sequenceTip, err := s.bridgeOutClient.GetAssetsUnlockedSequenceTip(ctx)
+	if err != nil {
+		return fmt.Errorf(
+			"failed to fetch assets unlocked sequence tip: [%w]",
+			err,
+		)
+	}
+
+	// This should never happen. Check just in case.
+	if sequenceTip.LT(s.lastAssetsUnlockedSequence) {
+		return fmt.Errorf(
+			"current AssetsUnlock sequence tip on Mezo lower than "+
+				"the last processed unlock sequence (%s vs %s)",
+			sequenceTip.String(),
+			s.lastAssetsUnlockedSequence.String(),
+		)
+	}
+
+	// If the unlock sequence tip has not advanced on Mezo return early.
+	if sequenceTip.Equal(s.lastAssetsUnlockedSequence) {
+		s.logger.Info(
+			"No new AssetsUnlocked events to fetch from Mezo",
+			"unlock_sequence_tip", s.lastAssetsUnlockedSequence.String(),
+		)
+		return nil
+	}
+
+	// The unlock sequence tip on Mezo is greater than the unlock sequence from
+	// the last fetched event. We need to fetch events from Mezo.
+	seqStart := s.lastAssetsUnlockedSequence.AddRaw(1)
+	seqEnd := sequenceTip.AddRaw(1)
+
+	// TODO: Do we need to split the request into batches?
+	//       Fetching new events is run frequently - there shouldn't be
+	//       many events to fetch.
+	events, err := s.bridgeOutClient.GetAssetsUnlockedEvents(
+		ctx,
+		seqStart,
+		seqEnd,
+	)
+	if err != nil {
+		return fmt.Errorf(
+			"failed to fetch AssetsUnlocked events [%s, %s): %w",
+			seqStart.String(),
+			seqEnd.String(),
+			err,
+		)
+	}
+
+	expectedLength := seqEnd.Sub(seqStart).Int64()
+	if int64(len(events)) != expectedLength {
+		return fmt.Errorf(
+			"fetched unexpected number of AssetsUnlocked events for "+
+				"range [%s, %s); expected %d, got %d",
+			seqStart.String(),
+			seqEnd.String(),
+			expectedLength,
+			len(events),
+		)
+	}
```

Comment:
Same comment as in https://github.com/mezo-org/mezod/pull/506#discussion_r2257570132